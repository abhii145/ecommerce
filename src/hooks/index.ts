import axios from "axios"
import {
  keepPreviousData,
  useQuery,
  UseQueryResult,
} from "@tanstack/react-query"
import { Category, Product, UseProductsParams, ProductResponse } from "../types"

const API_BASE_URL = "https://dummyjson.com/products"

export const useCategories = (): UseQueryResult<Category[]> => {
  return useQuery({
    queryKey: ["categories"],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/categories`)
      return response.data
    },
    staleTime: 1000 * 60 * 5,
    placeholderData: keepPreviousData,
  })
}

export const useProducts = ({
  limit,
  skip,
  searchQuery,
  category,
  sortBy,
  order,
}: UseProductsParams): UseQueryResult<ProductResponse> => {
  return useQuery({
    queryKey: ["products", limit, skip, searchQuery, category, sortBy, order],
    queryFn: async () => {
      let url = `${API_BASE_URL}?limit=${limit}&skip=${skip}&sortBy=${sortBy}&order=${order}`

      if (searchQuery) {
        url = `${API_BASE_URL}/search?q=${searchQuery}&limit=${limit}&skip=${skip}`
      } else if (category && category !== "All") {
        url = `${API_BASE_URL}/category/${category}?limit=${limit}&skip=${skip}&sortBy=${sortBy}&order=${order}`
      }

      const response = await axios.get(url)
      return response.data
    },
    placeholderData: keepPreviousData,
    staleTime: 1000 * 60 * 5,
  })
}

export const useSingleProduct = (id: string): UseQueryResult<Product> => {
  return useQuery({
    queryKey: ["singleProducts", id],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/${id}`)
      return response.data
    },
    staleTime: 1000 * 60 * 5,
  })
}

export const useSearchSuggestions = (query: string): UseQueryResult<Product[]> => {
  return useQuery({
    queryKey: ["searchSuggestions", query],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/search?q=${query}`)
      return response.data.products.map((product: Product) => ({
        ...product,
        category: product.category || "Uncategorized",
      }))
    },
    enabled: !!query,
    staleTime: 1000 * 60 * 5,
  })
}
