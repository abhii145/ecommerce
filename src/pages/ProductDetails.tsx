import { useParams } from "react-router"
import { useState, useEffect } from "react"
import { useDispatch, useSelector } from "react-redux"
import { useForm } from "react-hook-form"
import { z } from "zod"
import { zodResolver } from "@hookform/resolvers/zod"
import { addToCart } from "../store/cartSlice"
import { addToFavorites } from "../store/favoritesSlice"
import { addRecentlyViewed } from "../store/recentlyViewedSlice"
import { RootState } from "../store/store"
import { Favorite, Product, AddToCartProduct } from "../types"
import { toast } from "react-toastify"
import { useSingleProduct } from "../hooks"
import Loader from "../components/Loader"
import { renderStars } from "../utils/renderStars"
import { motion } from "framer-motion"
import { ProductCard } from "../components"

const pincodeSchema = z.object({
  pincode: z.string().regex(/^\d{6}$/, "Pincode must be exactly 6 digits"),
})

const ProductDetails = () => {
  const { id } = useParams<{ id: string }>()
  const [mainImage, setMainImage] = useState<string>("")
  const [isPincodeValid, setIsPincodeValid] = useState<boolean>(false)
  const dispatch = useDispatch()
  const favorites = useSelector((state: RootState) => state.favorites.items)
  const recentlyViewed = useSelector((state: RootState) => state.recentlyViewed.items)

  const { isLoading, data: product } = useSingleProduct(id)

  useEffect(() => {
    if (product?.images && product.images.length > 0) {
      setMainImage(product.thumbnail)
    }
  }, [product])

  useEffect(() => {
    if (product) {
      dispatch(addRecentlyViewed(product))
    }
  }, [product, dispatch])

  const handleImageClick = (image: string) => {
    setMainImage(image)
  }

  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm<{ pincode: string }>({
    resolver: zodResolver(pincodeSchema),
  })

  const handlePincodeChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value.replace(/\D/g, "")
    setValue("pincode", value)
    if (value.length === 6) {
      setIsPincodeValid(/^\d{6}$/.test(value))
    } else {
      setIsPincodeValid(false)
    }
  }

  const addToCartHandler = async (product: Product) => {
    const { id, title, price, thumbnail, brand, quantity, category } = product
    const addToCartProduct: AddToCartProduct = {
      id,
      title,
      price,
      thumbnail,
      brand: brand ?? "Locally produced",
      quantity,
      category,
      images: [],
      description: "",
      discountPercentage: 0,
      rating: 0,
      stock: 0,
      warrantyInformation: "",
      availabilityStatus: "",
      reviews: [],
    }
    await dispatch(addToCart(addToCartProduct))
    toast.success("Product added to cart", {
      position: "bottom-right",
      autoClose: 800,
    })
  }

  const addToFavoritesHandler = async (product: Product) => {
    const { id, title, price, thumbnail, description, rating } = product
    const isFavorite = favorites?.some(
      (item: Favorite) => item?.id === product.id
    )
    if (isFavorite) {
      toast.info("Product already in favorites", {
        position: "bottom-right",
        autoClose: 800,
      })
    } else {
      await dispatch(
        addToFavorites({ id, title, price, thumbnail, description, rating })
      )
      toast.success("Product added to favorites", {
        position: "bottom-right",
        autoClose: 800,
      })
    }
  }

  const filteredRecentlyViewed = recentlyViewed.filter(item => item.id !== product?.id)

  if (isLoading) {
    return <Loader />
  }

  return (
    <div className="bg-white-50 min-h-screen py-10">
      <div className="container mx-auto bg-white p-8 rounded-lg">
        <div className="grid grid-cols-1 md:grid-cols-2 gap-8">
          {/* Left: Images Section */}
          <motion.div
            className="flex flex-col items-center md:items-start space-y-4"
            initial={{ opacity: 0, y: 20 }}
            animate={{ opacity: 1, y: 0 }}
            transition={{ duration: 0.5 }}
          >
            <div className="relative w-full md:w-96 h-96 rounded-lg overflow-hidden shadow-lg">
              <img
                src={mainImage}
                alt={product.title}
                className="w-full h-full object-contain"
                loading="lazy"
              />
            </div>
            <div className="flex space-x-2 overflow-x-auto">
              {product?.images.map((image, index) => (
                <motion.img
                  key={index}
                  src={image}
                  alt={`${product.title} thumbnail ${index}`}
                  className={`w-16 h-16 object-cover cursor-pointer rounded-lg shadow-md ${
                    mainImage === image
                      ? "ring-2 ring-blue-500"
                      : "ring-1 ring-gray-300"
                  }`}
                  onClick={() => handleImageClick(image)}
                  whileHover={{ scale: 1.1 }}
                />
              ))}
            </div>
          </motion.div>

          {/* Right: Product Details */}
          <motion.div
            className="space-y-6"
            initial={{ opacity: 0, y: 20 }}
            animate={{ opacity: 1, y: 0 }}
            transition={{ duration: 0.5, delay: 0.2 }}
          >
            {/* Basic Info */}
            <h1 className="text-4xl font-bold text-gray-800">
              {product.title}
            </h1>
            <p className="text-gray-600 text-lg">{product.description}</p>

            {/* Price and Discount */}
            <div className="flex items-center space-x-4">
              <span className="text-2xl font-bold text-blue-600">
                ₹{product.price}
              </span>
              <span className="text-sm text-gray-500 line-through">
                ₹
                {(
                  product.price +
                  (product.price * product.discountPercentage) / 100
                ).toFixed(2)}
              </span>
              <span className="text-sm text-green-500 font-medium">
                ({product.discountPercentage}% OFF)
              </span>
            </div>

            {/* Pincode Input */}
            <form onSubmit={handleSubmit(() => {})} noValidate>
              <div className="flex flex-col lg:flex-row items-center space-y-4 lg:space-y-0 lg:space-x-4 mt-4">
                <div className="relative z-0">
                  <input
                    id="pincode"
                    type="text"
                    {...register("pincode")}
                    placeholder=" "
                    autoComplete="off"
                    aria-invalid={errors.pincode ? "true" : "false"}
                    aria-describedby="pincode-error"
                    className={`peer block py-2.5 px-0 text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none ${
                      errors.pincode ? "border-red-500" : ""
                    }`}
                    onChange={handlePincodeChange}
                    maxLength={6}
                  />
                  <label
                    htmlFor="pincode"
                    className={`absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 ${
                      errors.pincode ? "text-red-500" : ""
                    }`}
                    style={{ zIndex: 1 }}
                  >
                    Pincode
                  </label>
                  {errors.pincode && (
                    <p id="pincode-error" className="text-red-500 text-sm">
                      {errors.pincode?.message}
                    </p>
                  )}
                </div>
                {isPincodeValid && (
                  <div className="text-green-600">
                    <div className="flex items-center space-x-2">
                      <span>{product.shippingInformation}</span>
                    </div>
                  </div>
                )}
              </div>
            </form>

            {/* Additional Information */}
            <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
              <div className="flex items-center space-x-2">
                <span className="font-medium text-gray-700">Brand:</span>
                <span>{product.brand || "Locally produced"}</span>
              </div>
              <div className="flex items-center space-x-2">
                <span className="font-medium text-gray-700">Availability:</span>
                <span
                  className={
                    product.availabilityStatus === "Out of Stock"
                      ? "text-red-500 font-semibold"
                      : "text-green-500 font-semibold"
                  }
                >
                  {product.availabilityStatus}
                </span>
              </div>
              <div className="flex items-center space-x-2">
                <span className="font-medium text-gray-700">Warranty:</span>
                <span>
                  {product.warrantyInformation || "No warranty available"}
                </span>
              </div>
              <div className="flex items-center space-x-2">
                <span className="font-medium text-gray-700">Category:</span>
                <span>{product.category}</span>
              </div>
              <div className="flex items-center space-x-2">
                <span className="font-medium text-gray-700">Returns:</span>
                <span>
                  {product.returnPolicy || "30-day return policy available"}
                </span>
              </div>
            </div>

            {/* Actions */}
            <div className="flex space-x-4 mt-6">
              <button
                className={`bg-blue-600 text-white py-3 px-8 rounded-lg shadow-md flex items-center justify-center space-x-2 ${
                  product.availabilityStatus === "Out of Stock"
                    ? "opacity-50 cursor-not-allowed"
                    : "hover:bg-blue-700"
                }`}
                onClick={() => addToCartHandler(product)}
                disabled={product.availabilityStatus === "Out of Stock"}
              >
                <span>Add to Cart</span>
              </button>
              <button
                className="bg-red-500 text-white py-3 px-8 rounded-lg shadow-md flex items-center justify-center space-x-2 hover:bg-red-600"
                onClick={() => addToFavoritesHandler(product)}
              >
                <span>Add to Favorites</span>
              </button>
            </div>
          </motion.div>
        </div>

        {/* Recently Viewed Products */}
        {filteredRecentlyViewed.length > 0 && (
          <motion.div
            className="mt-16"
            initial={{ opacity: 0, y: 20 }}
            animate={{ opacity: 1, y: 0 }}
            transition={{ duration: 0.5, delay: 0.4 }}
          >
            <h2 className="text-2xl font-semibold text-gray-800 mb-4">
              Recently Viewed Products
            </h2>
            <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-4 gap-4">
              {filteredRecentlyViewed.map((item) => (
                <ProductCard key={item.id} {...item} />
              ))}
            </div>
          </motion.div>
        )}

        {/* Reviews */}
        <motion.div
          className="mt-16"
          initial={{ opacity: 0, y: 20 }}
          animate={{ opacity: 1, y: 0 }}
          transition={{ duration: 0.5, delay: 0.4 }}
        >
          <h2 className="text-3xl font-semibold text-gray-800 mb-4">
            Customer Reviews
          </h2>
          <div className="flex space-x-6 overflow-x-auto">
            {product.reviews.map((review, index) => (
              <motion.div
                key={index}
                className="bg-white p-6 rounded-lg shadow-md w-80 transition-transform duration-300 hover:scale-105"
                whileHover={{ scale: 1.05 }}
              >
                <div className="flex items-center space-x-2 mb-2">
                  <span className="font-semibold text-gray-800">
                    {review.reviewerName}
                  </span>
                  <span className="text-sm text-gray-500">
                    - {new Date(review.date).toLocaleDateString()}
                  </span>
                </div>
                <p className="text-gray-600 text-sm">{review.comment}</p>
                <div className="mt-2 flex items-center space-x-1">
                  <span className="font-medium text-gray-700">Rating:</span>
                  <div className="flex">{renderStars(review.rating)}</div>
                </div>
                <div className="mt-2 text-sm text-gray-500">
                  <span>Email: {review.reviewerEmail}</span>
                </div>
              </motion.div>
            ))}
          </div>
        </motion.div>
      </div>
    </div>
  )
}

export default ProductDetails
