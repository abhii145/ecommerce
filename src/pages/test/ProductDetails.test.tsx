/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, screen, fireEvent } from "@testing-library/react"
import { Provider } from "react-redux"
import { BrowserRouter as Router } from "react-router"
import { ToastContainer } from "react-toastify"
import configureStore from "redux-mock-store"
import ProductDetails from "../ProductDetails"
import { useSingleProduct } from "../../hooks"

jest.mock("../../hooks")
jest.mock("../../store/cartSlice", () => ({
  addToCart: jest.fn().mockReturnValue({ type: "cart/addToCart" }),
}))
jest.mock("../../store/favoritesSlice", () => ({
  addToFavorites: jest
    .fn()
    .mockReturnValue({ type: "favorites/addToFavorites" }),
}))

const mockStore = configureStore([])

describe("ProductDetails", () => {
  let store: any

  beforeEach(() => {
    store = mockStore({
      favorites: { items: [] },
      recentlyViewed: { items: [] }, // Ensure recentlyViewed is initialized
    })
  })

  const product = {
    id: "1",
    title: "Test Product",
    description: "Test Description",
    price: 100,
    discountPercentage: 10,
    rating: 4.5,
    brand: "Test Brand",
    warrantyInformation: "1 year",
    availabilityStatus: "In Stock",
    thumbnail: "test-thumbnail.jpg",
    images: ["image1.jpg", "image2.jpg"],
    shippingInformation: "Ships in 2 days",
    reviews: [
      {
        reviewerName: "John Doe",
        date: "2023-01-01",
        comment: "Great product!",
        rating: 5,
        reviewerEmail: "john.doe@example.com",
      },
    ],
  }

  it("snapshot", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it("renders product details", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
          <ToastContainer />
        </Router>
      </Provider>
    )
    expect(screen.getByText("Test Product")).toBeInTheDocument()
    expect(screen.getByText("Test Description")).toBeInTheDocument()
    expect(screen.getByText("₹100")).toBeInTheDocument()
    expect(screen.getByText("₹110.00")).toBeInTheDocument()
    expect(screen.getByText("(10% OFF)")).toBeInTheDocument()
    expect(screen.getByText("Brand:")).toBeInTheDocument()
    expect(screen.getByText("Test Brand")).toBeInTheDocument()
    expect(screen.getByText("Warranty:")).toBeInTheDocument()
    expect(screen.getByText("1 year")).toBeInTheDocument()
    expect(screen.getByText("Availability:")).toBeInTheDocument()
    expect(screen.getByText("In Stock")).toBeInTheDocument()
  })

  it("handles image click", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    const image = screen.getByAltText("Test Product thumbnail 0")
    fireEvent.click(image)
    expect(screen.getByAltText("Test Product")).toHaveAttribute(
      "src",
      "image1.jpg"
    )
  })

  it("renders customer reviews", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    expect(screen.getByText("Customer Reviews")).toBeInTheDocument()
    expect(screen.getByText("John Doe")).toBeInTheDocument()
    expect(screen.getByText("Great product!")).toBeInTheDocument()
    expect(screen.getByText("Rating:")).toBeInTheDocument()
    expect(screen.getByText("Email: john.doe@example.com")).toBeInTheDocument()
  })

  it("renders product images", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    product.images.forEach((image, index) => {
      expect(
        screen.getByAltText(`Test Product thumbnail ${index}`)
      ).toHaveAttribute("src", image)
    })
  })

  it("handles pincode change", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    const pincodeInput = screen.getByLabelText("Pincode")
    fireEvent.change(pincodeInput, { target: { value: "123456" } })
    expect(pincodeInput).toHaveValue("123456")
  })

  it("adds product to cart", async () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
          <ToastContainer />
        </Router>
      </Provider>
    )
    const button = screen.getByText("Add to Cart")
    fireEvent.click(button)
    expect(await screen.findByText("Product added to cart")).toBeInTheDocument()
  })

  it("adds product to favorites", async () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
          <ToastContainer />
        </Router>
      </Provider>
    )
    const button = screen.getByText("Add to Favorites")
    fireEvent.click(button)
    expect(
      await screen.findByText("Product added to favorites")
    ).toBeInTheDocument()
  })

  it("shows loader when loading", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: true,
      data: null,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    expect(screen.getByText("Loading...")).toBeInTheDocument()
  })

  it("displays shipping information when pincode is valid", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    const pincodeInput = screen.getByLabelText("Pincode")
    fireEvent.change(pincodeInput, { target: { value: "123456" } })
    expect(screen.getByText(product.shippingInformation)).toBeInTheDocument()
  })

  it("sets isPincodeValid to false when pincode is less than 6 digits", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    const pincodeInput = screen.getByLabelText("Pincode")
    fireEvent.change(pincodeInput, { target: { value: "123" } })
    expect(pincodeInput).toHaveValue("123")
    expect(
      screen.queryByText(product.shippingInformation)
    ).not.toBeInTheDocument()
  })

  it("handles add to cart with out of stock product", async () => {
    const outOfStockProduct = { ...product, availabilityStatus: "Out of Stock" }
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: outOfStockProduct,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
          <ToastContainer />
        </Router>
      </Provider>
    )
    const button = screen.getByText("Add to Cart")
    fireEvent.click(button)
    expect(
      await screen.queryByText("Product added to cart")
    ).not.toBeInTheDocument()
  })

  it("handles add to favorites with out of stock product", async () => {
    const outOfStockProduct = { ...product, availabilityStatus: "Out of Stock" }
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: outOfStockProduct,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
          <ToastContainer />
        </Router>
      </Provider>
    )
    const button = screen.getByText("Add to Favorites")
    fireEvent.click(button)
    expect(
      await screen.queryByText("Product added to favorites")
    ).not.toBeInTheDocument()
  })

  it("handles invalid pincode input", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    const pincodeInput = screen.getByLabelText("Pincode")
    fireEvent.change(pincodeInput, { target: { value: "abc123" } })
    expect(pincodeInput).toHaveValue("123")
    expect(
      screen.queryByText(product.shippingInformation)
    ).not.toBeInTheDocument()
  })

  it("handles valid pincode input", () => {
    ;(useSingleProduct as jest.Mock).mockReturnValue({
      isLoading: false,
      data: product,
    })
    render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    )
    const pincodeInput = screen.getByLabelText("Pincode")
    fireEvent.change(pincodeInput, { target: { value: "123456" } })
    expect(pincodeInput).toHaveValue("123456")
    expect(screen.getByText(product.shippingInformation)).toBeInTheDocument()
  })
})
