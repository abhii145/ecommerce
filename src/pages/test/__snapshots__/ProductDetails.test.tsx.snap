// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white-50 min-h-screen py-10"
  >
    <div
      class="container mx-auto bg-white p-8 rounded-lg"
    >
      <div
        class="grid grid-cols-1 md:grid-cols-2 gap-8"
      >
        <div
          class="flex flex-col items-center md:items-start space-y-4"
          style="opacity: 0; transform: translateY(20px);"
        >
          <div
            class="relative w-full md:w-96 h-96 rounded-lg overflow-hidden shadow-lg"
          >
            <img
              alt="Test Product"
              class="w-full h-full object-contain"
              loading="lazy"
              src="test-thumbnail.jpg"
            />
          </div>
          <div
            class="flex space-x-2 overflow-x-auto"
          >
            <img
              alt="Test Product thumbnail 0"
              class="w-16 h-16 object-cover cursor-pointer rounded-lg shadow-md ring-1 ring-gray-300"
              src="image1.jpg"
            />
            <img
              alt="Test Product thumbnail 1"
              class="w-16 h-16 object-cover cursor-pointer rounded-lg shadow-md ring-1 ring-gray-300"
              src="image2.jpg"
            />
          </div>
        </div>
        <div
          class="space-y-6"
          style="opacity: 0; transform: translateY(20px);"
        >
          <h1
            class="text-4xl font-bold text-gray-800"
          >
            Test Product
          </h1>
          <p
            class="text-gray-600 text-lg"
          >
            Test Description
          </p>
          <div
            class="flex items-center space-x-4"
          >
            <span
              class="text-2xl font-bold text-blue-600"
            >
              ₹100
            </span>
            <span
              class="text-sm text-gray-500 line-through"
            >
              ₹110.00
            </span>
            <span
              class="text-sm text-green-500 font-medium"
            >
              (10% OFF)
            </span>
          </div>
          <form
            novalidate=""
          >
            <div
              class="flex flex-col lg:flex-row items-center space-y-4 lg:space-y-0 lg:space-x-4 mt-4"
            >
              <div
                class="relative z-0"
              >
                <input
                  aria-describedby="pincode-error"
                  aria-invalid="false"
                  autocomplete="off"
                  class="peer block py-2.5 px-0 text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none focus:outline-none "
                  id="pincode"
                  maxlength="6"
                  name="pincode"
                  placeholder=" "
                  type="text"
                />
                <label
                  class="absolute text-sm text-gray-500 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6 "
                  for="pincode"
                  style="z-index: 1;"
                >
                  Pincode
                </label>
              </div>
            </div>
          </form>
          <div
            class="grid grid-cols-1 md:grid-cols-2 gap-4"
          >
            <div
              class="flex items-center space-x-2"
            >
              <span
                class="font-medium text-gray-700"
              >
                Brand:
              </span>
              <span>
                Test Brand
              </span>
            </div>
            <div
              class="flex items-center space-x-2"
            >
              <span
                class="font-medium text-gray-700"
              >
                Availability:
              </span>
              <span
                class="text-green-500 font-semibold"
              >
                In Stock
              </span>
            </div>
            <div
              class="flex items-center space-x-2"
            >
              <span
                class="font-medium text-gray-700"
              >
                Warranty:
              </span>
              <span>
                1 year
              </span>
            </div>
            <div
              class="flex items-center space-x-2"
            >
              <span
                class="font-medium text-gray-700"
              >
                Category:
              </span>
              <span />
            </div>
            <div
              class="flex items-center space-x-2"
            >
              <span
                class="font-medium text-gray-700"
              >
                Returns:
              </span>
              <span>
                30-day return policy available
              </span>
            </div>
          </div>
          <div
            class="flex space-x-4 mt-6"
          >
            <button
              class="bg-blue-600 text-white py-3 px-8 rounded-lg shadow-md flex items-center justify-center space-x-2 hover:bg-blue-700"
            >
              <span>
                Add to Cart
              </span>
            </button>
            <button
              class="bg-red-500 text-white py-3 px-8 rounded-lg shadow-md flex items-center justify-center space-x-2 hover:bg-red-600"
            >
              <span>
                Add to Favorites
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="mt-16"
        style="opacity: 0; transform: translateY(20px);"
      >
        <h2
          class="text-3xl font-semibold text-gray-800 mb-4"
        >
          Customer Reviews
        </h2>
        <div
          class="flex space-x-6 overflow-x-auto"
        >
          <div
            class="bg-white p-6 rounded-lg shadow-md w-80 transition-transform duration-300 hover:scale-105"
          >
            <div
              class="flex items-center space-x-2 mb-2"
            >
              <span
                class="font-semibold text-gray-800"
              >
                John Doe
              </span>
              <span
                class="text-sm text-gray-500"
              >
                - 1/1/2023
              </span>
            </div>
            <p
              class="text-gray-600 text-sm"
            >
              Great product!
            </p>
            <div
              class="mt-2 flex items-center space-x-1"
            >
              <span
                class="font-medium text-gray-700"
              >
                Rating:
              </span>
              <div
                class="flex"
              >
                <div
                  class="flex"
                >
                  <svg
                    class="text-yellow-500"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
                    />
                  </svg>
                  <svg
                    class="text-yellow-500"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
                    />
                  </svg>
                  <svg
                    class="text-yellow-500"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
                    />
                  </svg>
                  <svg
                    class="text-yellow-500"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
                    />
                  </svg>
                  <svg
                    class="text-yellow-500"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="mt-2 text-sm text-gray-500"
            >
              <span>
                Email: john.doe@example.com
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
